okta/request_throttle_transport.go

```diff
@@ -9,78 +9,90 @@ import (
 	"time"
 )
 
-var (
-	apiV1AppsEndpointCalls              = 0
-	apiV1AppsEndpointRateLimitLimit     = 0
-	apiV1AppsEndpointRateLimitResetTime time.Time
-	apiV1AppsEndpointMux                sync.Mutex
-	apiV1AppsPattern                    = regexp.MustCompile(
-		`/api/v1/apps$`)
-	apiV1AppsGroupsPattern = regexp.MustCompile(
-		`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/groups$`)
-	apiV1AppsUsersPattern = regexp.MustCompile(
-		`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/users$`)
-	apiV1AppsGroupPattern = regexp.MustCompile(
-		`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/groups/(?P<GroupID>[[:alnum:]]+)$`)
-)
-
-func isApiV1AppsEndpoint(path string) bool {
-	return len(apiV1AppsPattern.FindStringSubmatch(path))+
-		len(apiV1AppsGroupsPattern.FindStringSubmatch(path))+
-		len(apiV1AppsUsersPattern.FindStringSubmatch(path))+
-		len(apiV1AppsGroupPattern.FindStringSubmatch(path)) != 0
+func isAPIV1AppsEndpoint(path string) bool {
+	endpointsPatterns := []*regexp.Regexp{
+		// apps pattern
+		regexp.MustCompile(`/api/v1/apps$`),
+		// groups pattern
+		regexp.MustCompile(`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/groups$`),
+		// group pattern
+		regexp.MustCompile(`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/groups/(?P<GroupID>[[:alnum:]]+)$`),
+		// users pattern
+		regexp.MustCompile(`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/users$`),
+	}
+	// Check if endpoint match to one of the patterns.
+	var foundLen int
+	for _, p := range endpointsPatterns {
+		foundLen += len(p.FindStringSubmatch(path))
+	}
+	return foundLen != 0
 }
 
-type RequestThrottleTransport struct {
-	base        http.RoundTripper
-	maxRequests int
+type requestThrottleTransport struct {
+	base                                http.RoundTripper
+	percentageOfLimitRate               int
+	apiV1AppsEndpointCalls              int
+	apiV1AppsEndpointRateLimit          int
+	apiV1AppsEndpointRateLimitResetTime time.Time
+	sync.Mutex
 }
 
-func (throttleTransport *RequestThrottleTransport) RoundTrip(req *http.Request) (res *http.Response, err error) {
-	appsEndpoint := isApiV1AppsEndpoint(req.URL.Path)
-	if appsEndpoint {
-		log.Println("[DEBUG] special isApiV1AppsEndpoint request throttle handling")
-		apiV1AppsEndpointMux.Lock()
-		apiV1AppsEndpointCalls += 1
-		// TODO hardcoded 10 for testing, change it to calculating on the fly using
-		// apiV1AppsEndpointRateLimitLimit * c.maxRequests / 100
-		if apiV1AppsEndpointCalls >= 10 {
-			apiV1AppsEndpointCalls = 0
-			// add an extra margin to account for the clock skew
-			timeToSleep := apiV1AppsEndpointRateLimitResetTime.Add(2 * time.Second).Sub(time.Now())
-			if timeToSleep > 0 {
-				log.Printf(
-					"[INFO] Throttling /api/v1/apps requests, sleeping until rate limit reset for %s",
-					timeToSleep,
-				)
-				time.Sleep(timeToSleep)
-			}
-		}
-		apiV1AppsEndpointMux.Unlock()
+func (tt *requestThrottleTransport) RoundTrip(req *http.Request) (res *http.Response, err error) {
+	// Nothing to throttle, just bypass.
+	if !isAPIV1AppsEndpoint(req.URL.Path) {
```

Comment:
It worries me that this function could grow a lot when you add new endpoints. I think we should aim for something like this:
```
func (throttleTransport *RequestThrottleTransport) RoundTrip(req *http.Request) (res *http.Response, err error) {
	for i, _ := range throttleTransport.throttledEndpoints {
		throttleTransport.throttledEndpoints[i].preRequestHook(req.URL.Path)
	}
	resp, err := throttleTransport.base.RoundTrip(req)
	for i, _ := range throttleTransport.throttledEndpoints {
		throttleTransport.throttledEndpoints[i].postRequestHook(req.URL.Path)
	}
	return resp, err
}
```

I'd create a new struct that represents one rate limit:
```
type rateLimitThrottle struct {
	endpointsPatterns []*regexp.Regexp
	endpointCalls int
	rateLimit int
	rateLimitResetTime time.Time
	sync.Mutex
}
func (throttle *rateLimitThrottle) isPath(path string) bool {
	...
}
func (throttle *rateLimitThrottle) preRequestHook(path string) {
	...
}
func (throttle *rateLimitThrottle) postRequestHook(path string) {
	...
}
```
And do the `RequestThrottleTransport.throttledEndpoints` initialization in `NewRequestThrottleTransport`:
```
func NewRequestThrottleTransport(base http.RoundTripper, maxRequests int) *RequestThrottleTransport {
	apiV1AppsEndpoint := newRateLimitThrottle([]string{
		`/api/v1/apps$`,
		`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/groups$`,
		`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/groups/(?P<GroupID>[[:alnum:]]+)$`,
		`/api/v1/apps/(?P<AppID>[[:alnum:]]+)/users$`,
	})
	return &RequestThrottleTransport{
		base: base,
		maxRequests: maxRequests,
		throttledEndpoints: []rateLimitThrottle{apiV1AppsEndpoint},
	}
}
```